### ISSUE TITLE
[no-use-before-define] Add an option to ignore enums

### ISSUE THREAD

**Issue by Torhal** [typescript-eslint/typescript-eslint#1235]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-use-before-define": [
        "error",
        {
            classes: true,
            functions: false,
            typedefs: false,
        },
    ]
  }
}
```

```TS
// At bottom of file
enum MyEnum { /* ... */ }
```

**Expected Result**
No error, since enums are hoisted.

**Actual Result**
'MyEnum' was used before it was defined. eslint(@typescript-eslint/no-use-before-define)

**Additional Info**
I'm working around this by moving the enums to the top of the file, but I'd rather not do that as I prefer the code to be front-and-center.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.8.0` |
| `@typescript-eslint/parser`        | `2.8.0` |
| `TypeScript`                       | `3.7.2` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `12.13.0` |
| `npm`                              | `6.13.0` |

**Comment by bradzacher** [comment]
Hmm - looking into this more, you mention that "enums are hoisted", but they aren't actually.

```ts
const x = Foo.bar; // ts error: Enum 'Foo' used before its declaration

enum Foo { bar = 1 }
```
transpiles to
```js
const x = Foo.bar; // Runtime error: Uncaught TypeError: Cannot read property 'bar' of undefined
var Foo;
(function (Foo) {
    Foo[Foo["bar"] = 1] = "bar";
})(Foo || (Foo = {}));
```

in comparison to something that _is_ hoisted (functions):
```ts
const x = Foo();

function Foo() { }
```
this has no TS errors and has no runtime errors.

Do you have an example of where enums are hoisted?
Otherwise, I think that this is probably an inherently unsafe option that shouldn't exist.

**Comment by Torhal** [comment]
I actually meant the latter. I've never had occasion where I'd do something like in your example at the top level of a file.

I've had locally-used enums at the bottom of files in my codebases for at least two years now and was able to use them within functions above their declaration/definition without issue until I migrated from TSLint to ESLint with the @typescript-eslint rules.

As a workaround, since I'd rather keep the rule in place, I've moved them to the tops of the files they're in.

**Comment by bradzacher** [comment]
Cool - to clarify for you, things that are hoisted have their declarations moved to the top of the scope automatically.
Varaibles have their _definitions_ hoisted (not the declaration, just the definition which states "this variable exists in this scope").
```ts
function foo() {
  var y = x; // no runtime error
  var x = 1;

  const b = a; // runtime error - a used _before initialisation_
  const a = 1;

  const c = d; // runtime error - x is _not defined_
}
```

Function _declaration_s (`function foo() {}`) are completely hoisted:
```ts
const x = foo();  // no error, as the entire declaration an implementation of foo was hoisted
function foo() { return 1 }
```

Your specific use case is actually just how closures work - when you define a function it "closes over" the parent scope, meaning it tracks all of the references:

```ts
function MyFunc() {
  return MyEnum.A;
}

MyFunc(); // runtime error - Cannot read property 'A' of undefined

enum MyEnum {
  A = 1;
}

MyFunc(); // No error
```
In the above example, MyFunc closes over the reference to MyEnum, but nothing is hoisted, which is why it is a runtime error to call MyFunc before declaring MyEnum.

---

All that rambling aside...

If anything, I feel like this should just be handled as part of the `variables` option, because enums are treated exactly the same as a variable.
But it kind of does make sense to have a different option, because enums can be used as types.
The attached pr still makes sense.
